### ISSUE TITLE
`NeMoModelCheckpoint.on_train_end()` saves the epoch info incorrectly

### ISSUE THREAD

**Issue by yaox12** [NVIDIA/NeMo#6734]
**Describe the bug**

`NeMoModelCheckpoint.on_train_end()` saves the epoch info incorrectly, which would break resuming training from the last checkpoint.

For example, when the maximum time limit is reached on a cluster, we would usually save checkpoints and launch another job to resume training. Assume that the break happens in epoch 0, ideally, the epoch info of the fit loop should be like the following:
```python
In [3]: ckpt['loops']['fit_loop']['epoch_progress']
Out[3]:
{'total': {'ready': 1, 'completed': 0, 'started': 1, 'processed': 0},
 'current': {'ready': 1, 'completed': 0, 'started': 1, 'processed': 0}}
```
Then in the second job, the PTL trainer knows this is a restarting job. 

However, `NeMoModelCheckpoint` save the `-last.ckpt` in `on_train_end()`, where the `completed` and `processed` counters have been increased. So the epoch info is

```python
In [4]: ckpt_last['loops']['fit_loop']['epoch_progress']
Out[4]:
{'total': {'ready': 1, 'completed': 1, 'started': 1, 'processed': 1},
 'current': {'ready': 1, 'completed': 1, 'started': 1, 'processed': 1}}
```

This bug would break resuming training from the last checkpoint in several cases:
1. When `trainer.max_epochs = 1`, PTL thinks the max epoch has been reached and then stops training immediately.
2. When `trainer.max_epochs > 1`, PTL can restart training but starts from epoch 1 instead of 0, and calculate the total number of batches incorrectly. The logs look like:
```
# at the end of the first job.
Epoch 0:  39%|███▉      | 411/1055 [3:43:23<5:50:01, 32.61s/it, loss=1.25, v_num=jxb0, reduced_train_loss=1.180, global_step=390.0, consumed_samples=49920.0, val_loss=1.140]
wandb: Waiting for W&B process to finish... (success).

# at the beginning of the second job.
Training: 391it [00:00, ?it/s]
Training:   0%|          | 0/645 [00:00<00:00, -18221920.71it/s]
Epoch 1:   0%|          | 0/645 [00:00<?, ?it/s]                
Epoch 1:  64%|██████▍   | 412/645 [00:36<00:20, 11.43it/s]
Epoch 1:  64%|██████▍   | 412/645 [00:36<00:20, 11.42it/s, loss=1.13, v_num=25nj, reduced_train_loss=1.130, global_step=391.0, consumed_samples=49920.0]
```
As we can see, it restarts training from Epoch 1, and the total number of batches decreases from 1055 to 645. Ideally, it should restart at Epoch 0, step 412/1055.

**Steps/Code to reproduce bug**

We run SFT using `examples/nlp/language_modeling/tuning/conf/megatron_gpt_sft.yaml`.

**Comment by athitten** [comment]
@yaox12 it seems like the epoch info getting saved incorrectly in the checkpoint is because of `_save_last_checkpoint` being done upon train end in the callback (aka hook) `on_train_end()`. In this callback PTL is incrementing all the counters like trainer.current_epoch and others to the next epoch . This issue is not seen when  `_save_last_checkpoint` is done in other callbacks like `on_train_batch_end` or `on_validation_end` or `on_train_epoch_end`. We are currently working on supporting PTL 2.0 with NeMo, will check if this is fixed by PTL in 2.0.

**Comment by BestJuly** [comment]
Hi @athitten, do you know how to understand the iteration index here? Is it also caused by the PTL callback?
The regular steps in one epoch should be 1055 in this case, but if we restart training similar to the log, it seems that the first epoch would only last for 645 steps. And when it comes to the next epoch in such case, the total steps become 645.

If considering the following example: 10 samples and 10 iterations (bsz=1), train 2 epoch in total, 4 iterations per run due to time limit which need to launch training for 5 times. I have no idea which samples to be used in the training.
```
# 10 samples: [A B C D E F G H I J]

# 1st run, use sample [A B C D]
Epoch 0: 0/10 # start
Epoch 0: 4/10 # end

# 2nd run, use which sample?
Epoch 1: 4/6 # start, epoch index becomes 1 because of the issue
Epoch 2: 2/6 # end

# 3rd run, use which sample?
Epoch 3: 2/4 # start, epoch index becomes 3 because of the issue
Epoch 4: 2/4 # end
...
```

If the training dataset will not be shuffled for each epoch, then some samples may be used for multiple times and some would never be used, would this happen in current case?
